### PR TITLE
chore: improve rate of false OK sends on websockets

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This PR comprises two small changes. First, we try to return errors from writes on websockets to the caller more promptly by returning the error from `s.conn.WriteMessage` rather than returning success always. Second, we buffer failed sends when the error was not ours (validation), assuming the socket will be closed immediately and the recovery routine will kick off. When finished, we resend the buffered messages.

## Test Plan
- [ ] run all upstream CI a few times, try to write a test.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: agent/internal/agent.go [hunk 1]
```diff
@@ -52,7 +52,8 @@ type agent struct {
 	masterProto  string
 	masterClient *http.Client
 
-	reconnecting bool
+	reconnecting         bool
+	reconnectSendBacklog []interface{}
```
Comment (ioga) [inline agent/internal/agent.go:56]: maybe `[]api.WriteMessage`?

Comment (ioga) [inline agent/internal/agent.go:56]: I wish we could have a "reliability" wrapper around the sockets so all the backlog retry logic was hidden away in there, and not be an intrinsic part of `agent` actor here.

File: master/pkg/actor/api/ws.go [hunk 2]
```diff
@@ -32,6 +33,16 @@ const (
 	MaxWebsocketMessageSize = 128 * 1024 * 1024
 )
 
+// MaxMaxWebsocketMessageSizeExceededError indicates to the call that the message exceeded the
+// maximum websocket size.
+type MaxMaxWebsocketMessageSizeExceededError struct {
```
Comment (ioga) [inline master/pkg/actor/api/ws.go:38]:
```suggestion
type MaxWebsocketMessageSizeExceededError struct {
```

File: agent/internal/agent.go (hunk 1 above)
Comment (stoksc) [reply agent/internal/agent.go:56]: that is my aim in a follow up pr, but this one needs to make release it seems

File: agent/internal/agent.go [hunk 3]
```diff
@@ -351,6 +377,13 @@ func (a *agent) attemptReconnect(ctx *actor.Context) bool {
 	for i := 0; i < aproto.AgentReconnectAttempts; i++ {
 		switch err := a.connect(ctx); {
 		case err == nil:
+			for msg := range a.reconnectSendBacklog {
+				if err = ctx.Ask(a.socket, msg).Error(); err != nil {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply agent/internal/agent.go:381]: we try to retry once, once the connection has been reestablished. if we fail again immediately, we just call it quits and fail.

File: agent/internal/agent.go [hunk 4]
```diff
@@ -89,14 +90,34 @@ func (a *agent) Receive(ctx *actor.Context) error {
...
+			switch err := ctx.Ask(a.socket, wsm).Error(); err.(type) {
+			case nil:
+				return nil
+			case api.MaxMaxWebsocketMessageSizeExceededError:
+				ctx.Log().WithError(err).Error("container log too large")
+			default:
+				ctx.Log().WithError(err).Warnf("writing container log %v", msg)
+				a.bufferSendForRecovery(ctx, wsm)
+			}
+		} else {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply agent/internal/agent.go:119]: there is a way to run an agent standalone for testing, i think that's all this is.

**Comment by stoksc** [comment]
@ioga going to just close this

**Pull request closed without merging**